### DISCUSSION TITLE
Passing a list of values and generate individual HelmRelease files

### DISCUSSION THREAD

**Discussion by DevopsCareer** [fluxcd/flux2#3165] [Q&A]
Hello,
I have a override value in my HelmRelease and it can take five possible values, instead of creating 5 different HelmRelease files is there a way I can pass those as a list and FluxCD generate 5 such HelmRelease files?
eg I have :
```
spec
  values
     APP_NAME: "app"
     ID: 
      - "000"
      - "111"
      - "222"
      - "333"
      - "444"
```

 I want on the run time (when we deploy HelmRelease) on cluster to interpret `ID` list and substitute the list values to create 5 HelmRelease.yamls. If this is not possible what is recommended way to handle list values in HelmRelease ?

**Comment by kingdonb** [comment]
There is no way that I'm aware of to make this expression as elegant in the way that you're describing, you need a configuration language where you can do looping constructs if you want to maintain this value set as a list and generate HelmReleases from it. Maybe cue lang or jsonnet?

Kubernetes resources are declarative not imperative, so they aren't usually programmable in this way, but you can do that kind of imperative loop in some intent definition in your repo, that then unfolds via configuration delivery pipeline (CI) into expanded manifests that Flux can apply from the generated form.

There are lots of *functional* ways that Kustomize can do config generation, but not what you described, if I read it correctly what you are asking, and that's the reason why. If you needed to get those values into 5 HelmReleases, you can use one base HelmRelease and build five configmaps with the five different values, you can use configmap generators (or just write out the configmaps) for this, with `valuesFrom`

The added benefit of a configMapGenerator is that your HelmRelease spec is updated every time the configmap changes, so you don't have to wait `spec.interval` because the change in secret actually results in Kustomize making a change inside of the spec of the HelmRelease, which results in a new generation and immediate notification of the Helm Controller that it's time to reconcile this HelmRelease that has a change in it.

You can use either `valuesFrom` or environment substitution in the HelmRelease definition to get a value from the parent Kustomization into the HelmRelease. But there isn't any list form that you can use that I'm aware of. Maybe you can pass the list of strings in a one-line array, like this (don't quote me, I'm an engineerTM):

```
  ID: [ "000", "111", "222", "333", "444"]
```

The Flux string substitution probably doesn't let you do this that I'm aware of, and it didn't really get us any closer to answering your top question, but maybe if you need this there's a chance it will work somewhere. But what definitely won't work is a multiline substitution.

Here's some supportive documentation:

https://fluxcd.io/flux/guides/helmreleases/#refer-to-values-in-configmaps-generated-with-kustomize

And just in case the values are secret, and you can't leave them in a plaintext YAML file for configmap generator to consume, if you have a tenant where decryption is available (SOPS) then you can always encrypt them there in-place with SOPS

We added this guide recently:

https://fluxcd.io/flux/guides/helmreleases/#refer-to-values-in-secret-generated-with-kustomize-and-sops

Hope this helps! Thanks for posting here
